src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,286 @@
+/*!
+ * Copyright 2019 by XGBoost Contributors
+ */
+#include <thrust/functional.h>
+#include <thrust/random.h>
+#include <thrust/transform.h>
+#include <thrust/transform_reduce.h>
+#include <xgboost/host_device_vector.h>
+#include <xgboost/logging.h>
+
+#include <algorithm>
+
+#include "../../common/compressed_iterator.h"
+#include "../../common/random.h"
+#include "gradient_based_sampler.cuh"
+
+namespace xgboost {
+namespace tree {
+
+GradientBasedSampler::GradientBasedSampler(EllpackPageImpl* page,
+                                           size_t n_rows,
+                                           BatchParam batch_param,
+                                           float subsample,
+                                           int sampling_method)
+    : original_page_(page),
+      batch_param_(batch_param),
+      is_external_memory_(page->matrix.n_rows != n_rows),
+      subsample_(subsample),
+      is_sampling_(subsample < 1.0),
+      sampling_method_(sampling_method),
+      sample_rows_(n_rows * subsample) {
+  monitor_.Init("gradient_based_sampler");
+
+  if (is_external_memory_) {
```

Comment:
If you have to have a bool and an if statement in every method, I think it suggests that you need two separate classes.

Reply:
I changed the in-memory implementation to match external memory. The only special handling now is when not sampling, in external memory mode we concatenate the pages together.